### ISSUE TITLE
more support for operation nodes with multiple args

### ISSUE THREAD

**Issue by evykassirer** [josdejong/mathjs#739]
If I create an operation node with list of 3 arguments

- `toString()` doesn't use the operator symbol anymore

e.g.
`var node1 = new math.expression.node.ConstantNode(2);`
`var node2 = new math.expression.node.ConstantNode(3);`
`var node3 = new math.expression.node.ConstantNode(4);`
`var sum = new math.expression.node.OperatorNode('+', 'add', [node1, node2, node3]);`

`sum.toString()` --> `"add(2, 3, 4)"`

I propose that this print "2 + 3 + 4" instead. I think toTex() has the same issue

- `eval()` doesn't evaluate the sum anymore

`sum.eval()` --> `VM221:48 Uncaught TypeError: Too many arguments in function add (expected: 2, actual: 3)`

I propose that since Operator nodes take a list of arguments and not exactly two arguments, they should be able to support eval() for any number of arguments in the arg list

**Comment by FSMaxB** [comment]
What is the usecase for this?

**Comment by evykassirer** [comment]
Partially it's so that if people create valid nodes they should be able to use them like any other.

The reason I need it is for a CAS step by step simplifier built off of mathJS that Jos is planning to use as an extension to mathjs. I need to merge arguments to simplify easily, and am using my own functions to print and eval, but if the functionality was part of mathJS that'd be awesome

**Comment by FSMaxB** [comment]
I'm not completely sure at this point what the implications for `toString` and `toTex` are regarding operator precedence. I will have to think about that in more detail.

**Comment by FSMaxB** [comment]
But for now I think it might be okay to just always wrap everything in parentheses.

**Comment by evykassirer** [comment]
I think it's the same as two arguments
Unary minus needs it's argument wrapped in parentheses (and it doesn't make sense for it to have multiple arguments)
Subtraction and division also don't make sense having more than two arguments because e.g. 2-3-4 could be 2-(3-4) or (2-3)-4, you can't "subtract 3 things together"

You can however add or multiply three things together. I could have been more clear that those are the two cases that I care about. They should keep parens around it if it would need it for two arguments, but won't need parens otherwise.

Does that sound right?

**Comment by FSMaxB** [comment]
Yes. Makes sense.

From a `toTex` and `toString` perspective this change seems quite feasible.

And as long as an operator is commutative it makes sense to support it in the underlying functions like `add` and `multiply` as well.

**Comment by josdejong** [comment]
Agree. I think it's nice to extend `add` and `multiply` to support more than two arguments. That will basically render `sum` and `prod` redundant which is ok (we could keep them as an alias). In the Nodes of the expression parser we only have to do a few changes in `toString` and `toTex`. We can keep the parser as is for now (parsing `2+3+4` as two nested additions). In the future we could make the parser smarter if we want. At least, this makes life easier for the step solver that @evykassirer is working on, and could be interesting for @tetslee too (see #734).

**Comment by FSMaxB** [comment]
I think I can do the `toString` and `toTex` part of this during the course of next week.

**Comment by josdejong** [comment]
:+1: 

I can adjust `add` and `multiply` within a few weeks (not sure whether I can make it this week).

**Comment by josdejong** [comment]
I've extended `add` and `multiply` to accept more than two arguments. Was more tricky than I thought, I first had to fix an old, tough bug in `typed-function` related to any type parameters.

**Comment by ericman314** [comment]
This will be very helpful. If I recall, I used nodes named `add` and `multiply` in my work for `simplify`, and I used more than two arguments to convert a nested tree of plus'es and times'es into a flat structure. However, I never attempted to evaluate those functions directly--they were only used for convenience. I always intended to convert back to a nested tree. But this will no doubt be useful.
